packages/flare/python/cron_job_ingest_events.py

```diff
@@ -29,46 +32,119 @@ def main():
         raise Exception(str(e))
 
     app: client.Application = splunk_service.apps[APP_NAME]
-    flare_api = FlareAPI(app=app)
+    api_key: Optional[str] = None
+    tenant_id: Optional[int] = None
+    for item in app.service.storage_passwords.list():
+        if item.content.username == "api_key":
+            api_key = item.clear_password
 
-    from_ = get_from_value(app=app)
-    event_feed = flare_api.retrieve_feed(from_=from_)
-    set_from_value(app=app, next=event_feed["next"])
+        if item.content.username == "tenant_id":
+            tenant_id = (
+                int(item.clear_password) if item.clear_password is not None else None
+            )
 
-    if event_feed["items"]:
-        for item in event_feed["items"]:
-            print(json.dumps(item))
+    if not api_key:
+        raise Exception("API key not found")
 
+    if not tenant_id:
+        raise Exception("Tenant ID not found")
+    flare_api = FlareAPI(app=app, api_key=api_key, tenant_id=tenant_id)
 
-def get_from_value(app: client.Application) -> Optional[str]:
+    from_ = get_from_value(app=app, tenant_id=tenant_id)
+    start_date = get_start_date_value(app=app)
+    while True:
+        event_feed = flare_api.retrieve_feed(from_=from_, start_date=start_date)
+        save_values(app=app, tenant_id=tenant_id, next=event_feed["next"])
```

Comment:
`save_state()`? 

At first glance, I'm thinking that this function is doing too many things. I'd be tempted to break it up into several other functions. If it were a case for atomicity, one function and a bunch of logic to ensure atomicity would make sense, but that's not what's happening here. 

What do you think?

Reply:
I split the methods in 3 separate methods instead, one for creating the collection, one to save the start date and the other to save the new next